### PR TITLE
Prepare for deployment in container

### PR DESCRIPTION
**Pull request by martijnjanssen**
Properties have been added in the build. Still not quite sure what port and what URL will be served on, I could use some help with that.

Testing this PR can be done by running:
- `./gradlew build`
- `docker build -t payments .`
- `docker run -e "SPRING_PROFILES_ACTIVE=production" -e "SPRING_MAIL_HOST=localhost" -e "SPRING_MAIL_PORT=1025" -e "PAYMENTS_MAIL_SENDER=" -e "SPRING_DATASOURCE_URL=jdbc:hsqldb:mem:payments" -e "SPRING_DATASOURCE_USERNAME=sa" -e "SPRING_DATASOURCE_PASSWORD=" -e "PAYMENTS_MOLLIEKEY=${FILL_MOLLIEKEY_HERE}" -e "PAYMENTS_ADMIN_GROUPS=chbeheer,bestuur,hoothub,${SOME_GROUP_YOU_ARE_IN}" -e "PAYMENTS_ADMIN_USERNAME=${SOME_ADMIN_USERNAME}" -e "PAYMENTS_ADMIN_PASSWORD=${SOME_ADMIN_PASSWORD}" -p 8080:8080 payments:latest`

### CODE REVIEW

File: src/main/resources/application-container.properties [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+endpoints.cors.allowed-origins=*
+endpoints.cors.allowed-methods=*
+endpoints.cors.allowed-headers=*
+endpoints.cors.max-age=3600
+security.basic.enabled=false
+payments.paymentReturnUrl=http://ch.tudelft.nl
+payments.molliekey=${PAYMENTS_MOLLIEKEY}
+payments.returnUrl="http://localhost:9000"
+connect.issuerUri=https://connect.ch.tudelft.nl/
+connect.clientUri=http://localhost:8080/
```
Comment (martijnjanssen) [inline src/main/resources/application-container.properties]: Do we want to change any of these URLs?

**Comment by praseodym** [comment]
Please just use a separate `production` or `prod` profile instead, we will then enable that profile in production.

File: Dockerfile [hunk 2]
```diff
@@ -3,4 +3,4 @@ ADD build/libs/payments.jar /srv/payments.jar
 WORKDIR /srv
 EXPOSE 8080
 RUN mkdir config
-ENTRYPOINT ["java","-Dspring.profiles.active=container","-jar","/srv/payments.jar"]
+ENTRYPOINT ["java","-Dspring.profiles.active=production","-jar","/srv/payments.jar"]
```
Comment (praseodym) [inline Dockerfile]: Not every instance of this container will always be running in production?

File: Dockerfile [hunk 3]
```diff
@@ -3,4 +3,4 @@ ADD build/libs/payments.jar /srv/payments.jar
 WORKDIR /srv
 EXPOSE 8080
 RUN mkdir config
```
Comment (praseodym) [inline Dockerfile]: Why is this needed?

File: Dockerfile (hunk 2 above)
Comment (martijnjanssen) [reply Dockerfile]: What is the correct way of enabling a profile then?

File: Dockerfile (hunk 3 above)
Comment (martijnjanssen) [inline Dockerfile]: This is the folder that contains our mailtemplates. Is it possible to map this directory to somewhere outside the container to keep the mailtemplates between multiple deployments?

Comment (martijnjanssen) [reply Dockerfile]: This is the folder where the `oidc-client-registration.json` will we written. We're not quite sure whether this is important. But if the folder is not present, the application crashes.

**Comment by martijnjanssen** [comment]
I could also create a gradle task to run a testing container, would this be useful to do besides adding a docker section to the README.md.

File: Dockerfile (hunk 2 above)
Comment (praseodym) [reply Dockerfile]: For example using the `SPRING_PROFILES_ACTIVE` environment variable (which you do now).

File: Dockerfile (hunk 3 above)
Comment (praseodym) [reply Dockerfile]: You [should be using a statically registered client in production](https://github.com/WISVCH/payments/blob/ed74371fa4387ab904beeac52974ce6c464f88b2/src/main/java/ch/wisv/payments/SecurityConfig.java#L178). For development I'd simply change the location where the file is written to.

**Comment by martijnjanssen** [comment]
@praseodym could you take a look at the `StaticClientConfiguration`? @skamoen and I can't figure out how it works/is supposed to work.

File: src/main/resources/application-production.properties [hunk 4]
```diff
@@ -0,0 +1,38 @@
+spring.mail.host=${SPRING_MAIL_HOST}
```
Comment (praseodym) [inline src/main/resources/application-production.properties]: Not needed

File: src/main/resources/application-production.properties [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+spring.mail.port=${SPRING_MAIL_PORT}
```
Comment (praseodym) [inline src/main/resources/application-production.properties]: Not needed

File: src/main/resources/application-production.properties [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+payments.mail.sender=${PAYMENTS_MAIL_SENDER}
```
Comment (praseodym) [inline src/main/resources/application-production.properties]: Not needed

File: src/main/resources/application-production.properties [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+spring.thymeleaf.check-template-location=true
+spring.thymeleaf.prefix=classpath:/templates/
+spring.thymeleaf.suffix=.html
+spring.thymeleaf.mode=HTML5
+spring.thymeleaf.encoding=UTF-8
+spring.thymeleaf.content-type=text/html
+
+##DATABASE PROPERTIES
+spring.datasource.driverClassName=org.hsqldb.jdbc.JDBCDriver
+spring.datasource.url=${SPRING_DATASOURCE_URL}
```
Comment (praseodym) [inline src/main/resources/application-production.properties]: Same for all other env variables

**Review by praseodym — CHANGES_REQUESTED** [review]

**Comment by praseodym** [comment]
Did you make any progress on the OIDC config with my suggestions from this morning?

**Comment by martijnjanssen** [comment]
No, I don't have any more free time right now and am unable to figure out how this works. Unfortunately I cant't fix this in the coming days, we really want to start deploying this since we have a deadline, could you please look at this?

File: src/main/resources/application-production.properties [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+spring.jpa.generate-ddl=true
+spring.jpa.hibernate.ddl-auto=validate
+spring.jpa.properties.hibernate.default_schema=PUBLIC
+endpoints.cors.allow-credentials=*
+endpoints.cors.allowed-origins=*
+endpoints.cors.allowed-methods=*
+endpoints.cors.allowed-headers=*
+endpoints.cors.max-age=3600
+security.basic.enabled=false
+payments.paymentReturnUrl=http://ch.tudelft.nl
```
Comment (praseodym) [inline src/main/resources/application-production.properties]: is dit echt de waarde die je deze setting wilt geven?

Comment (martijnjanssen) [reply src/main/resources/application-production.properties]: I think we do, since this is the fallbackUrl if no returnUrl is specified for a payment. We'll have to check all url functions though, since it's not quite clear what url is supposed to do what in the properties, it could definitely use some more clarity.

Comment (praseodym) [reply src/main/resources/application-production.properties]: Please use https then :-)

File: config/application-test.properties [hunk 9]
```diff
@@ -15,12 +15,13 @@ spring.jpa.hibernate.ddl-auto=create-drop
 spring.jpa.properties.hibernate.default_schema=PUBLIC
 spring.thymeleaf.mode=HTML
 security.basic.enabled=false
-a5l.paymentReturnUrl=url
-a5l.molliekey=key
-payments.returnUrl=url
+payments.returnUrl=http://localhost:9000
+payments.paymentReturnUrl=http://localhost:9000
```
Comment (praseodym) [inline config/application-test.properties]: what's the difference between these?

**Review by praseodym — APPROVED** [review]

**Comment by TimvdLippe** [comment]
I will do the honor of hitting the green button :heart: #whatcouldpossiblygowrong
